sw/host/tests/chip/gpio/src/sleep_pin_mio_dio_val.rs

```diff
@@ -0,0 +1,252 @@
+    false, // IOR5, MIO Pad 40
+    true,  // IOR6, MIO Pad 41
+    true,  // IOR7, MIO Pad 42
+    true,  // IOR10, MIO Pad 43
+    true,  // IOR11, MIO Pad 44
+    true,  // IOR12, MIO Pad 45
+    true,  // IOR13, MIO Pad 46
+];
+
+// IOA6, IOB6, IOC9-12, IOR5-7 and IOR10-13.
+const VALID_MIO_PADS_SILICON: &[bool] = &[
+    false, // IOA0, MIO Pad 0
+    false, // IOA1, MIO Pad 1
+    false, // IOA2, MIO Pad 2
+    false, // IOA3, MIO Pad 3
+    false, // IOA4, MIO Pad 4
+    false, // IOA5, MIO Pad 5
+    true,  // IOA6, MIO Pad 6
+    false, // IOA7, MIO Pad 7
+    false, // IOA8, MIO Pad 8
+    false, // IOB0, MIO Pad 9
+    false, // IOB1, MIO Pad 10
+    false, // IOB2, MIO Pad 11
+    false, // IOB3, MIO Pad 12
+    false, // IOB4, MIO Pad 13
+    false, // IOB5, MIO Pad 14
+    true,  // IOB6, MIO Pad 15
+    false, // IOB7, MIO Pad 16
+    false, // IOB8, MIO Pad 17
+    false, // IOB9, MIO Pad 18
+    false, // IOB10, MIO Pad 19
+    false, // IOB11, MIO Pad 20
+    false, // IOB12, MIO Pad 21
+    false, // IOC0, MIO Pad 22
+    false, // IOC1, MIO Pad 23
+    false, // IOC2, MIO Pad 24
+    false, // IOC3, MIO Pad 25
+    false, // IOC4, MIO Pad 26
+    false, // IOC5, MIO Pad 27
+    false, // IOC6, MIO Pad 28
+    false, // IOC7, MIO Pad 29
+    false, // IOC8, MIO Pad 30
+    true,  // IOC9, MIO Pad 31
+    true,  // IOC10, MIO Pad 32
+    true,  // IOC11, MIO Pad 33
+    true,  // IOC12, MIO Pad 34
+    false, // IOR0, MIO Pad 35
+    false, // IOR1, MIO Pad 36
+    false, // IOR2, MIO Pad 37
+    false, // IOR3, MIO Pad 38
+    false, // IOR4, MIO Pad 39
+    true,  // IOR5, MIO Pad 40
+    true,  // IOR6, MIO Pad 41
+    true,  // IOR7, MIO Pad 42
+    true,  // IOR10, MIO Pad 43
+    true,  // IOR11, MIO Pad 44
+    true,  // IOR12, MIO Pad 45
+    true,  // IOR13, MIO Pad 46
+];
+
+#[derive(Debug, Parser)]
+struct Opts {
+    #[command(flatten)]
+    init: InitializeTest,
+
+    /// Console receive timeout.
+    #[arg(long, value_parser = humantime::parse_duration, default_value = "600s")]
+    timeout: Duration,
+}
+
+fn get_valid_mio_pads(interface: &str) -> &'static [bool] {
+    match interface {
+        "hyper310" => VALID_MIO_PADS_FPGA_CW310,
+        "teacup" => VALID_MIO_PADS_SILICON,
+        _ => &[],
+    }
+}
+
+fn sleep_pin_mio_dio_test(opts: &Opts, transport: &TransportWrapper) -> Result<()> {
+    let interface = opts.init.backend_opts.interface.as_str();
+    let uart = transport.uart("console")?;
+    log::info!("Starting host side");
+
+    let mut expected_pin_values: [u8; 47] = [0; 47];
+    for _ in 0..get_valid_mio_pads(interface).len() {
+        let msg = UartConsole::wait_for(&*uart, r"MIO \[(\d+)\]: (\d+)", opts.timeout)?;
+        let index: usize = msg[1].parse()?;
+        expected_pin_values[index] = msg[2].parse()?;
+    }
+
+    UartConsole::wait_for(&*uart, r"Going to sleep", opts.timeout)?;
+
+    for (index, valid) in get_valid_mio_pads(interface).iter().enumerate() {
+        if *valid {
```

Comment:
Unsure if this is actually clearer or not, so feel free to ignore

```suggestion
    for index in get_valid_mio_pads(interface)
        .iter()
        .enumerate()
        .filter_map(|(i, valid)| valid.then_some(i))
    {
```